proptest-state-machine/src/test_runner.rs

```diff
@@ -0,0 +1,195 @@
+    }
+
+    /// Run the test sequentially. You typically don't need to override this
+    /// method.
+    fn test_sequential(
+        config: Config,
+        mut ref_state: <Self::Reference as ReferenceStateMachine>::State,
+        transitions: Vec<
+            <Self::Reference as ReferenceStateMachine>::Transition,
+        >,
+    ) {
+        let trans_len = transitions.len();
+        #[cfg(feature = "std")]
+        if config.verbose >= super::INFO_LOG {
+            eprintln!();
+            eprintln!("Running a test case with {} transitions.", trans_len);
+        }
+        #[cfg(not(feature = "std"))]
+        let _ = (config, trans_len);
+
+        let mut concrete_state = Self::init_test(&ref_state);
+
+        // Check the invariants on the initial state
+        Self::check_invariants(&concrete_state, &ref_state);
+
+        for (ix, transition) in transitions.into_iter().enumerate() {
+            #[cfg(feature = "std")]
+            if config.verbose >= super::INFO_LOG {
+                eprintln!();
+                eprintln!(
+                    "Applying transition {}/{}: {:?}",
+                    ix + 1,
+                    trans_len,
+                    transition
+                );
+            }
+            #[cfg(not(feature = "std"))]
+            let _ = ix;
+
+            // Apply the transition on the states
+            ref_state = <Self::Reference as ReferenceStateMachine>::apply(
+                ref_state,
+                &transition,
+            );
+            concrete_state =
+                Self::apply(concrete_state, &ref_state, transition);
+
+            // Check the invariants after the transition is applied
+            Self::check_invariants(&concrete_state, &ref_state);
+        }
+
+        Self::teardown(concrete_state)
+    }
+}
+
+/// This macro helps to turn a state machine test implementation into a runnable
+/// test. The macro expects a function header whose arguments follow a special
+/// syntax rules: First, we declare if we want to apply the state machine
+/// transitions sequentially or concurrently (currently, only the `sequential`
+/// is supported). Next, we give a range of how many transitions to generate,
+/// followed by `=>` and finally, an identifier that must implement
+/// `StateMachineTest`.
+///
+/// ## Example
+///
+/// ```rust,ignore
+/// struct MyTest;
+///
+/// impl StateMachineTest for MyTest {}
+///
+/// prop_state_machine! {
+///     #[test]
+///     fn run_with_macro(sequential 1..20 => MyTest);
+/// }
+/// ```
+///
+/// This example will expand to:
+///
+/// ```rust,ignore
+/// struct MyTest;
+///
+/// impl StateMachineTest for MyTest {}
+///
+/// proptest! {
+///     #[test]
+///     fn run_with_macro(
+///         (initial_state, transitions) in MyTest::sequential_strategy(1..20)
+///     ) {
+///        MyTest::test_sequential(initial_state, transitions)
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! prop_state_machine {
```

Comment:
Not really an issue for this PR, but this macro makes me glad that we're not shipping this in `proptest` proper just yet - I have a feeling it will be possible to get a nicer API with the attr-macro

Reply:
good point, I'm looking forward to replacing the decl macros! :pray: